src/libraries/Microsoft.PowerFx.Connectors/Execution/HttpFunctionInvoker.cs

```diff
@@ -160,33 +163,56 @@ public async Task<FormulaValue> DecodeResponseAsync(HttpResponseMessage response
         {
             // $$$ Do we need to check response media type to confirm that the content is indeed json?
             var json = await response.Content.ReadAsStringAsync();
+            var statusCode = (int)response.StatusCode;
+            var message = statusCode == 200 ? string.Empty : $"Connector call failed ({response.StatusCode}), request {response.RequestMessage.RequestUri}, request headers {FormatHeaders(response.RequestMessage.Headers)}";
 
-            if (!response.IsSuccessStatusCode)
+            return statusCode switch
             {
-                var msg = $"Connector call failed {response.StatusCode}): " + json;
+                int i when i >= 300 && i < 400 => FormulaValue.NewError(new ExpressionError() { Kind = ErrorKind.ConnectorWarning, Severity = ErrorSeverity.Warning, Message = message }),
+                int i when i >= 400 => FormulaValue.NewError(new ExpressionError() { Kind = ErrorKind.ConnectorError, Severity = ErrorSeverity.Critical, Message = message }),
+                
+                // 1xx, 2xx
+                _ => string.IsNullOrWhiteSpace(json) ? FormulaValue.NewBlank(_returnType) : FormulaValue.FromJson(json)
+            };
+        }
 
-                // $$$ Do any connectors have 40x behavior here in their response code?
-                // or 201 long-ops behavior?
+        private static string FormatHeaders(HttpRequestHeaders headers)
+        {
+            var sb = new StringBuilder(1024);
+            var b = false;
 
-                // $$$ Still type this. 
-                return FormulaValue.NewError(new ExpressionError
+            foreach (var header in headers)
+            {
+                // We don't want to log authentication tokens
+                if (header.Key == "Authorization")
                 {
-                    Kind = ErrorKind.Unknown,
-                    Message = msg
-                });
-            }
+                    continue;
```

Comment:
> continue; [](http://example.com/codeflow?start=19&length=10)


Are there other secret headers? It may be best to just not log them. What does Power Apps do?

Reply:
They don't provide any debug info AFAIK
I think we can do a better job
I don't know of any of secret but we can use a while list if you prefer to be safe